# manifest: PT-033
# base: 868d8d9216da361c666519652115e23688db5211
# mechanism: capability
# expiry: TBD
# tests: TBD
# owner: TBD
# seeded: 2026-08-30 whole-file diff of the fork vs upstream; P5 shrinks/splits this (0 feature sub-entries)
--- a/transformer_engine/pytorch/ops/basic/layer_norm.py
+++ b/transformer_engine/pytorch/ops/basic/layer_norm.py
@@ -1,3 +1,5 @@
+# This file was modified for portability to AMDGPU
+# Copyright (c) 2025-2026, Advanced Micro Devices, Inc. All rights reserved.
 # Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 #
 # See LICENSE for license information.
@@ -13,6 +15,9 @@
 import torch
 
 from transformer_engine_torch import layernorm_bwd, layernorm_fwd
+from torch.utils.cpp_extension import IS_HIP_EXTENSION
+if IS_HIP_EXTENSION:
+    from ...triton_kernels.norms_common import te_layernorm_fwd_triton, te_layernorm_bwd_triton
 from ...constants import TE_DType
 from ...cpu_offload import is_cpu_offload_enabled, mark_activation_offload
 from ...export import is_in_onnx_export_mode
@@ -202,7 +207,9 @@
 
         # Compute layer norm
         sm_margin = self._sm_margins["forward" if ctx.requires_grad else "inference"]
-        y, means, rstdevs = layernorm_fwd(
+        use_layernorm_triton = bool( int(os.environ.get('NVTE_USE_LAYERNORM_TRITON', '0')) ) and IS_HIP_EXTENSION
+        layernorm_fwd_func = te_layernorm_fwd_triton if use_layernorm_triton else layernorm_fwd
+        y, means, rstdevs = layernorm_fwd_func(
             x,
             w,
             b,
@@ -244,7 +251,9 @@
         w = maybe_dequantize(self.weight, dtype).view((inner_dim,))
 
         # Compute layer norm backward pass
-        dx, dw, db = layernorm_bwd(
+        use_layernorm_triton = bool( int(os.environ.get('NVTE_USE_LAYERNORM_TRITON', '0')) ) and IS_HIP_EXTENSION
+        layernorm_bwd_func = te_layernorm_bwd_triton if use_layernorm_triton else layernorm_bwd
+        dx, dw, db = layernorm_bwd_func(
             dy,
             x,
             means,
